source/debian/dat2pgm.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH SPYVIEW SECTION "May  6, 2003"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
dat2pgm
.SH SYNOPSIS
.B dat2pgm [-l] [-h] [-p value (50)] < infile.dat > outfile.pgm
.br
.SH DESCRIPTION
.PP
A command line utility that converts an ascii data file into a pgm image.
.PP
As input, it will read a whitespace separated matrix of floats. The
program will determine the extrema of the data, and then map the data
onto 50% of the output image dynamic range, centered at half of the
output maxval of 65535. The output is then written as a 16 bit pgm
file that includes comment fields that are read by spyview that
specify the mapping of the grey value on the the floating point data
values in the input file.
.PP
The program can also produce and output image that is given by the
(base 10) log of the absolute value of input data, by specifying the
-l command line option.
.PP
The default percentage of the output dynamic range used is 50%, in
order to allow maximum flexibility in saturating the colormap in
spyview. The percentage of the output dynamic range to use can be
specified using the -p option.
.PP
The -h option shows a breif usage message.
.PP
.SH SEE ALSO
.BR spyview3d (1),
.BR spyview (1),
.BR spybrowse (1)
.br
.SH AUTHOR
This manual page was written by Gary Steele <gsteele@electron.mit.edu>,
for the Debian project (but may be used by others).







debian/changelog
spyview (0.1-38) local; urgency=low

  * Serious changes since the last update!

 -- gsteele <gsteele@newkavli.nano.tudelft.nl>  Sat, 18 Aug 2007 11:02:30 -0400

spyview (0.1-37) local; urgency=low

  * Added loads of stuff, including the sweet new postscript output, and 
    an updated man page.

 --  <gsteele@qtmartini.sqt.tn.tudelft.nl>  Fri, 24 Nov 2006 16:43:46 +0100

spyview (0.1-36) local; urgency=low

  * Minor update to man page, very significant update to the software...

 --  <gsteele@qtmartini.sqt.tn.tudelft.nl>  Wed, 25 Oct 2006 09:55:37 +0200

spyview (0.1-35) local; urgency=low

  * Too many new features to list...including severely outdated man pages...

 -- odie <gsteele@qtmartini.sqt.tn.tudelft.nl>  Fri, 25 Aug 2006 11:58:50 +0200

spyview (0.1-34) local; urgency=low

  * Added pdf files and a lot of new features.

 -- odie <odie@electron.mit.edu>  Fri, 19 May 2006 05:43:36 -0400

spyview (0.1-33) local; urgency=low

  * Included stminfo in the distribution.

 --  <gsteele@vortex.ashoori.mit.edu>  Mon, 25 Apr 2005 16:27:49 -0400

spyview (0.1-32) local; urgency=low

  * Added support for a 3D binary floating point format that I call "mtx"
  * Added HSV rotation wheel for adjusting colors of a colormap
  * Added a new blue-green-pink colormap designed specifically for 
    my incompressible strip simulation data, as well as a new transfer 
    function adjustment that centers the colormap on the peak of the 
    histogram
  * Added a column-by-column subtraction routine to the image processing.

 --  <gsteele@vortex.ashoori.mit.edu>  Sat, 22 Jan 2005 15:53:15 -0500

spyview (0.1-31) local; urgency=low

  * UI enhancements, added support for arbitrary sized colormaps
  * Parameters are now saved in PPM file as comments
  * Modified matlabpink colormap so that it does not have a pre-applied 
    gamma correction

 --  <gsteele@vortex.ashoori.mit.edu>  Sun, 21 Nov 2004 17:47:10 -0500

spyview (0.1-30) local; urgency=low

  * Improved histogram plot

 --  <gsteele@vortex.ashoori.mit.edu>  Tue,  2 Nov 2004 10:36:04 -0500

spyview (0.1-29) local; urgency=low

  * Big enhancements to spybrowse: added a new Fl_Listbox widget

 --  <gsteele@vortex.ashoori.mit.edu>  Tue,  2 Nov 2004 08:49:57 -0500

spyview (0.1-28) local; urgency=low

  * Added support for separate x and y zooms, updated man page.

 --  <gsteele@vortex.ashoori.mit.edu>  Thu,  5 Aug 2004 16:45:20 -0400

spyview (0.1-27) local; urgency=low

  * Increased line buffer size of dat2pgm to 1024*1024 bytes. Also 
    fixed the logarithm so that it correctly output the base10 exponent.

 --  <gsteele@vortex.ashoori.mit.edu>  Wed,  4 Aug 2004 16:35:58 -0400

spyview (0.1-26) local; urgency=low

  * Removed ugly colormaps again after they snuck back in.

 --  <gsteele@vortex.ashoori.mit.edu>  Tue, 27 Jul 2004 12:52:10 -0400

spyview (0.1-25) local; urgency=low

  * Added a program dat2pgm for converting matrix ascii files into 
    pgm files that include the spyview comment fields for the axis
    scaling.

 --  <gsteele@vortex.ashoori.mit.edu>  Wed, 21 Jul 2004 21:09:09 -0400

spyview (0.1-24) local; urgency=low

  * Added a bunch of stuff, including arbitrary line cuts, support for
    changing x,y,z scaling/units of image using a dialog box, and other
    great stuff. Also, took topometrix colormaps out of distribution, since 
    they were just clogging things up and were mostly crap anyway.

 --  <gsteele@vortex.ashoori.mit.edu>  Fri,  2 Apr 2004 20:23:13 -0500

spyview (0.1-23) local; urgency=low

  * Added more normalize on load options to spyview.

 --  <gsteele@vortex.ashoori.mit.edu>  Tue,  2 Mar 2004 14:54:15 -0500

spyview (0.1-22) local; urgency=low

  * Added some colormaps, finally fixed filename stuff so we can load
    files that are in subdirectories.

 -- Gary Steele <gsteele@electron.mit.edu>  Tue, 20 Jan 2004 18:30:16 -0500

spyview (0.1-21) local; urgency=low

  * Fixed lbl subtraction.

 -- Gary Steele <gsteele@electron.mit.edu>  Sat, 10 Jan 2004 19:45:21 -0500

spyview (0.1-20) local; urgency=low

  * Added a new program spyrotate 
  * Some code cleanup

 -- Gary Steele <gsteele@vortex.ashoori.mit.edu>  Tue,  9 Dec 2003 14:37:12 -0500

spyview (0.1-19) local; urgency=low

  * Some bugfixes, changes shortcut 's' to save file as ppm. 

 -- Gary Steele <gsteele@vortex.ashoori.mit.edu>  Wed, 26 Nov 2003 10:50:48 -0500

spyview (0.1-18) local; urgency=low

  * Fixed a bug with the plane subtraction in the saveFile routine.

 -- Gary Steele <gsteele@vortex.ashoori.mit.edu>  Sun, 19 Oct 2003 13:33:22 -0400

spyview (0.1-17) local; urgency=low

  * Added an shlib.local file to override libc dependecy.

 -- Gary Steele <gsteele@electron.mit.edu>  Tue,  2 Sep 2003 15:15:50 -0400

spyview (0.1-16) local; urgency=low

  * Recompiled with gcc-3.3

 -- Gary Steele <gsteele@electron.mit.edu>  Tue,  2 Sep 2003 12:38:16 -0400

spyview (0.1-15) local; urgency=low

  * Fixed up gl2ps scaling, added rollers for colormap in spyview3d

 -- Gary Steele <gsteele@electron.mit.edu>  Sun,  3 Aug 2003 17:04:58 -0400

spyview (0.1-14) local; urgency=low

  * Changed gnuplot xsection commands so you can zoom with the mouse
    properly.

 -- Gary Steele <gsteele@electron.mit.edu>  Tue, 29 Jul 2003 15:09:35 -0400

spyview (0.1-13) local; urgency=low

  * Added plane subtraction to spyview3d. 

 -- Gary Steele <gsteele@electron.mit.edu>  Tue, 22 Jul 2003 22:29:31 -0400

spyview (0.1-12) local; urgency=low

  * Added basic colormap manipulation support to spyview3d.

 -- Gary Steele <gsteele@electron.mit.edu>  Tue, 22 Jul 2003 16:16:21 -0400

spyview (0.1-11) local; urgency=low

  * Added support for reading unit and scales out of AFM and Stm 
    files. The crosssections are now plots showing physical units.
    (The zaxis unit is ignored for Stm files, since it's usually
    still +/- 10V).
    
 -- Gary Steele <gsteele@electron.mit.edu>  Tue, 22 Jul 2003 14:15:09 -0400

spyview (0.1-10) local; urgency=low


source/spyview/dat2pgm.C
#include <stdio.h>
#include <math.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <assert.h>
#include <vector>
#include <string>
using namespace std;

#define LINESIZE 1024*1024

void usage()
{
  fprintf(stderr, "dat2pgm [-i{d,m,z,x}] [-l] [-v] [-m] [-x] [-p percent (50)] < infile.dat > outfile.pgm\n\n");
  fprintf(stderr, "    -l take log?\n");
  fprintf(stderr, "    -v verbose?\n");
  fprintf(stderr, "    -p percent of dynamic range to use in output image [50]\n");
  fprintf(stderr, "    -m generate a matlab ascii matrix file\n");
  fprintf(stderr, "    -x generate a spyview compatible mtx file instead\n");
  fprintf(stderr, "    -i Behaviour on incorrect columns (Drop line, fill with Mean, fill with Zero, [eXit])\n");

  exit(0);
}

int main(int argc, char **argv)
{
  int w, col;
  int h, row;
  int dolog = 0;
  int verbose = 0;
  typedef enum { DROP = 'd', FILL_MEAN = 'm', FILL_ZERO = 'z', EXIT = 'x'} incorrect_column_t;
  incorrect_column_t incorrect_column = EXIT;
  bool interp=false;
  bool matlab = false; // If set to true, output a text file appropriate for matlab instead of a pgm file.
  bool mtxout = false;

  char c;
  double percent = 50;
  
  typedef vector<string> comments_t;
  comments_t comments;
    
  while ((c = getopt(argc, argv, "lvhp:i:mxI")) != -1)
    {
      switch (c)
	{
        case 'I':
	  interp=true;
	  break;
	case 'i':
	  incorrect_column = static_cast<incorrect_column_t>(optarg[0]);
	  break;
	case '?':
	  usage();
	  break;
	case 'h':
	  usage();
	  break;
	case 'l':
	  dolog = 1;
	  break;
	case 'x':
	  mtxout = true;
	  break;
        case 'm':
          matlab = true;
	  break;
	case 'v':
	  verbose = 1;
	  break;
	case 'p':
	  if (sscanf(optarg, "%lf", &percent) != 1)
	    usage();
	  break;
	}
    }
  
  if (percent > 100) percent = 100;
  if (percent < 0) percent = 0;

  vector<double> data;

  char linebuffer[LINESIZE];
  char *p;

  double datamax = -1e100;
  double datamin = 1e100;
  double datatmp;

  double absdatamin = 1e100;
  
  char sep[] = " \t\n\r";

  row = col = w = h = 0;
  
  while (1)
    {
      if (fgets(linebuffer, LINESIZE, stdin) == NULL)
	break;
   
      // Figure out what kind of line this is
      for(p = linebuffer; *p != 0 && isspace(*p); p++)
	;
      if(*p == '#')
	{
	  comments.push_back(p);
	  continue;
	}
      else if(*p == '\n' || *p == 0)
	continue;

      col = 0;
      p = strtok(linebuffer, sep);
      while (p != NULL)
	{
	  if (sscanf(p, "%lf", &datatmp) != 1)
	    {
	      fprintf(stderr, "dat2pgm: invalid data at row %d col %d: \"%s\"\n", row, col, p);
	      exit(-1);
	    }
	  data.push_back(datatmp);
	  if (datatmp < datamin) datamin = datatmp;
	  if (datatmp > datamax) datamax = datatmp;
	  if (fabs(datatmp) < absdatamin && datatmp != 0.0) absdatamin = fabs(datatmp);
	  col++;
	  p = strtok(0, sep);
	}
      if (row == 0)
	w = col;
      else if (w != col)  // Col is the number of columns we have read, w is the number the other lines have had.
	{
	  if (col == 0)
	    break;      // ignore empty lines at the end.
	  if(col < w) // This row is too short
	    {
	      double d = 0; // This is used for ZERO; Mean updates the value so they can share insertion code.	 
	      switch(incorrect_column)
		{
		case EXIT:	      
		  fprintf(stderr, "number of columns %d at row %d does not match width of previous rows (%d)\n", col, row, w);
		  exit(-1);
		case DROP:
		  data.erase(data.begin()+(row*w),data.end());
		  row--;
		  break;
		case FILL_MEAN:
		  for(int i = 0; i < col; i++)
		    d += data[row*w + i];
		  d = d / col;
		case FILL_ZERO: // Warning! Falling case!
		  assert(col < w);
		  for(int i = 0; i < (w-col); i++)
		    data.push_back(d);
		  break;
		default:
		  fprintf(stderr, "Invalid argument for -i: %c\n",incorrect_column);
		  exit(-1);
		}
	    }
	  else // This row is too long
	    {
	      double d = 0; // This is used for ZERO; Mean updates the value so they can share insertion code.	 
	      switch(incorrect_column)
		{
		case EXIT:	      
		  fprintf(stderr, "number of columns %d at row %d does not match width of previous rows (%d)\n", col, row, w);
		  exit(-1);
		case FILL_MEAN:		  
		  fprintf(stderr, "number of columns %d at row %d does not match (too big, used to be %d); fill_mean is not ready yet for this case.\n",col,row,w);
		  fprintf(stderr, "Using fill_zero.");
		case FILL_ZERO: // Warning! Falling case!		  
		  {
		    fprintf(stderr, "number of columns %d at row %d does not match (too big, used to be %d)\n",col,row,w);
		    vector<double> dtmp;
		    dtmp.resize((row+1)*col);
		    for(int i = 0; i < row; i++)
		      {
			for(int j = 0; j < w; j++) // This is very inefficent, but easy to understand.
			  dtmp[i*col+j] = data[i*w+j];
			for(int j = w; j < col; j++)
			  dtmp[i*col+j] = 0.0;
		      }
		    for(int j = 0; j < col; j++) // Don't forget the last line.
		      dtmp[row*col + j] = data[row*w+j];
		    w = col;
		    data = dtmp;
		  }
		  break;
		case DROP:
		  data.erase(data.begin(),data.begin()+(row*w));
		  row = 0;
		  w = col;
		  break;
		default:
		  fprintf(stderr, "Invalid argument for -i: %c\n",incorrect_column);
		  exit(-1);
		}
	      
	    }
	  
	}
      row++;
      assert(data.size() == row * w);
    }
  h = row;

  if (verbose)
    fprintf(stderr, "%d x %d points\nMax = %e\nMin = %e\nAbsdatamin = %e\n", 
	    w, h, datamax, datamin, absdatamin);
  
  unsigned char tmp;
  
  if(interp)
    {
      double lastd=0;
      for(int i = 0; i < h*w; i++)
	if(isnan(data[i]))
	  {
	    data[i]=lastd; //(datamax+datamin)/2.0;
	    lastd=0;
	  }
	else
	  lastd=data[i];
    }
    else      
      for(int i = 0; i < h*w; i++)
	if(isnan(data[i]))
	  data[i]=0; //(datamax+datamin)/2.0;

  if (dolog)
    {
      datamax = -1e100;
      datamin = 1e100;
      for (int i=0; i<h; i++)
	{
	  for (int j=0; j<w; j++)
	    {  
	      data[i*w+j] = log10(fabs(data[i*w+j])+absdatamin);
	      if (data[i*w+j] < datamin) datamin = data[i*w+j];
	      if (data[i*w+j] > datamax) datamax = data[i*w+j];
	    }
	}
      if (verbose)
	fprintf(stderr, "LogMax = %g\nLogMin = %g\n", datamax, datamin);
    }

  if (mtxout)
    {
      fprintf(stdout, "1 %d %d 8\n", h, w);
      for (int i=0; i<h; i++)
	{
	  for(int j=w-1; j>-1; j--)
	    {
	      double t = data[i*w+j];
	      fwrite(&t, sizeof(t), 1, stdout);
	    }
	}
      return 0;
    }

  if(matlab) // This is not quite a null-op; we strip comments and fill out the lines to all be the same length.
    {
      for(int i = 0; i < h; i++)
	{
	  for(int j = 0; j < w; j++)
	    {
	      double t = data[i*w+j];
	    fprintf(stdout,"%.9g ",t);
	    }
	  fprintf(stdout,"\n");
	}
      return 0;
    }

  // Rescale data min and max to remap to the requested dynamic range 
  double center = (datamax+datamin)/2;
  double half_width  = (datamax-datamin)/2;

  datamin = center - half_width * 100 / percent;
  datamax = center + half_width * 100 / percent;

 
  fprintf(stdout, "P5\n%d %d\n#zmin %e\n#zmax %e\n",w,h, datamin, datamax);
  fprintf(stdout, "#xmin %d\n#xmax %d\n#ymin %d\n#ymax %d\n", 0, h, w, 0);
  fprintf(stdout, "#xunit X\n#yunit Y\n#zunit Z\n");

  for(comments_t::iterator i = comments.begin(); i != comments.end(); i++)
    fprintf(stdout,"%s",i->c_str());
  fprintf(stdout,"65535\n");

  for (int i=0; i<h; i++)
    {
      for (int j=0; j<w; j++)
	{
	  datatmp = (data[i*w+j] - datamin)/(datamax - datamin) * 65535.0;
	  //fprintf(stdout, "%.0f ", datatmp);
	  if (datatmp < 0.0 || datatmp > 65535.0 + 1e-6)
	    {
	      fprintf(stderr, "\n%d %d %f ", i,j, datatmp);
	    }
	  tmp = (unsigned char) ((int)datatmp / 256);
	  fwrite(&tmp, 1, 1, stdout);
	  tmp = (unsigned char) ((int)datatmp % 256);
	  fwrite(&tmp, 1, 1, stdout);
	}
      //fprintf(stdout,"\n");
    }
  
  if (verbose)
    fprintf(stderr, "Done.\n");
}
